Volume: Use ramp tokens and define the ramp tokens in all topologies

singalsu opened the pull request:
This PR consists of two commits. The first contains the changes for volume component. The second updates all topologies to define the volume ramp tokens. This approach does not need kernel update.

I removed the [RFC][Do not merge] labels since the code now seems to pass the CI tests.

singalsu commented:
@emilchudzik @zrombel This is the fix for issue #1297. Does the CI test set the volume new() IPC parameters initial_ramp and ramp? The parameter ramp should be set to 0 (SOF_VOLUME_LINEAR) and initial_ramp to e.g. 250 to preserve the current linear ramp slope. If initial_ramp is left to zero the volume jumps directly to new gain value and it could trigger an error in CI testst.

tlauda commented on src/audio/volume.c:
```diff
@@ -250,6 +264,61 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
+			inc = VOL_ZERO_DB;
+
+		delta = cd->tvolume[chan] - cd->volume[chan];
+		if (delta < 0)
```
Adding ABS macro would be useful.

singalsu replied:
Yep, will do.

tlauda commented on src/audio/volume.c:
```diff
@@ -250,6 +264,61 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
+
+		trace_volume("chan = %d", chan);
+		trace_volume("delta = %d", delta);
+		trace_volume("delta_abs = %d", delta_abs);
```
Can be done in one trace message.

singalsu replied:
These are something to remove. I was just checking that the code is sane with these.

singalsu replied:
These would be removed, or maybe converted to tracev().

tlauda commented on src/audio/volume.c:
```diff
@@ -250,6 +264,61 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
+
+		cd->ramp_increment[chan] = inc;
+
+		trace_volume("ramp_increment = %d", cd->ramp_increment[chan]);
```
Can be done in one trace message.

singalsu replied:
Also these are not needed. The t_us recompute is just for trace and can be removed also.

singalsu replied:
This could be tracev().

singalsu commented on src/audio/volume.c:
```diff
@@ -195,6 +194,9 @@ static struct comp_dev *volume_new(struct sof_ipc_comp *comp)
 		cd->tvolume[i] =  cd->volume[i];
 	}
 
+	trace_volume("initial_ramp = %d", vol->initial_ramp);
```
Oh, sorry, that's something I was about to remove. But on the other hand it could be useful to preserve and cleanup like you suggest.

singalsu commented on src/audio/volume.c:
```diff
@@ -222,8 +224,14 @@ static void volume_free(struct comp_dev *dev)
 	trace_volume("volume_params()");
 
+	trace_volume("initial_ramp = %d", vol->initial_ramp);
+	trace_volume("ramp = %d", vol->ramp);
```
The trace in params() is redundant, they are available in new().

zrombel commented:
@singalsu at this point in test we are using ramp=SOF_VOLUME_LINEAR and initial_ramp=0. I tested this PR manually and initial_ramp parameter seems to work correctly. We will set initial_ramp to 250 for current tests and prepare cases to test initial_ramp parameter. I see that at this point only SOF_VOLUME_LINEAR is supported. @singalsu do you plan to implement other ramp shapes in this PR?

singalsu commented:
@zrombel Good, thanks for adding ramp time in test! This PR will support only linear ramp. I'll enter an enhancement issue for logaritmic ramp and do it later. It's simple to add but will need exp() function into SOF math library. The zero crossings timed volume updates need more work (replace vol_work) but are possible and feasible also (avoid clicks when adjusting). Those could have another enhancement issue.

singalsu commented:
@tlauda I've now addressed your comments. I'm not totally sure why MIN() and MAX() are done like they are now but I followed the example with new ABS() function. I didn't add saturation to ABS. We could have a saturating inline function as well like abs_int32(). This macro is for generic use known not to wrap.

singalsu commented:
Thanks @tlauda @zrombel for the review effort! I'll change this to normal PR now.

singalsu commented:
The code is same as in previous PR version. I updated the commit title and improved a bit the commit body text.

ranj063 commented on src/audio/volume.h:
```diff
@@ -74,17 +74,17 @@
- * step in Q1.16 to use in vol_work function  is computed from the length.
+ * \brief Macro for volume ramp step computation
+ * Volume ramp step as Q1.16 is computed with equation
+ * step = 2^16 * 1000 / RAMP_UPDATE_US / ramp_length_ms .
```
@singalsu is this comment correct? where is ramp_length_ms in the macro below?

singalsu replied:
Yep, need to fix this comment.

ranj063 commented on src/audio/volume.c:
```diff
@@ -250,6 +257,52 @@ static inline void volume_set_chan(struct comp_dev *dev, int chan, uint32_t vol)
+		 * initial_ramp time. Calculate estimated ramp length [us].
+		 */
+		t_us = delta_abs * VOL_RAMP_UPDATE_US / inc;
+		if (t_us > pga->initial_ramp * 1000) {
```
@singalsu for my understanding when will this ever happen? the target volume is limited to VOL_MAX or VOL_MIN. so the delta can ever be maximum of VOL_ZERO_DB isnt it?

singalsu replied:
@ranj063 This happens pretty often when volume control is above 0 dB. This step re-calculate is the beef of the slow gain settling fix (while kernel does not provide the used min/max range). 

VOL_MAX is +42 dB so the the linear mute->unity gain equal transition is only e.g. 0 dB -> +6 dB jump. Such volume transitions can be made by pressing keys 0 - 9 in alsamixer ASCII graphics view.

wenqingfu commented:
does this need corresponding driver change? @singalsu

ranj063 commented:
> does this need corresponding driver change? @singalsu

@wenqingfu I dont think so. The volume tokens are already parsed by the driver.

lgirdwood commented:
@singalsu does min/max range require an ABI change ?

singalsu commented:
@lgirdwood There's no need for ABI change, the fields are just zeros now. I think I should implement in FW side the code to utilize the min/max when it will be made available by kernel driver. It's easy to make such test driver version with fixed values to test. For actual kernel work to extract the volume scale I've filed an enhancement issue.
